File: android/transport/src/main/kotlin/cc/bsns/ssh/transport/TerminalTransport.kt
package cc.bsns.ssh.transport

/**
 * What the terminal UI needs from a live connection, regardless of whether it's
 * an SSH PTY or a mosh UDP session. Both [SshSession] and [MoshSession] satisfy
 * it, so the terminal widget binds to the transport, not a concrete type — the
 * Android analogue of the iOS `TerminalTransport` boundary.
 */
interface TerminalTransport {
    var onOutput: ((ByteArray) -> Unit)?
    fun write(data: ByteArray)
    fun resize(cols: Int, rows: Int)
    fun close()
}

File: app/Sources/Session/TerminalTransport.swift
import Foundation
import BsnsSSHCore

/// The terminal byte-stream behind a `TerminalSession`, independent of how it's
/// carried — an interactive SSH channel (`SSHShell`) or a mosh UDP session
/// (`MoshSession`). The session binds to this, so swapping transports (or
/// reconnecting) never touches the view. Port forwarding is SSH-only and stays
/// off this protocol; `TerminalSession` reaches for `SSHShell` directly there.
protocol TerminalTransport: AnyObject {
    var onOutput: (@Sendable (ArraySlice<UInt8>) -> Void)? { get set }
    var onClosed: (@Sendable (String?) -> Void)? { get set }
    func write(_ bytes: ArraySlice<UInt8>)
    func resize(cols: Int32, rows: Int32)
    /// Ask the transport to replay its current terminal contents if it owns a
    /// separate framebuffer. Byte-stream transports can ignore this.
    func requestDisplayRefresh()
    func disconnect()
}

extension TerminalTransport {
    func requestDisplayRefresh() {}
}

// SSHShell already exposes exactly this surface.
extension SSHShell: TerminalTransport {}
